### ISSUE TITLE
SQLCipher cipher_page_size changed automatically.

### ISSUE THREAD

**Issue by Rozariozaro** [sqlcipher/sqlcipher#525]
### Expected Behavior
I am using current configuration settings 
cipher_salt="salt_key";
cipher_plaintext_header_size=32; 
journal_mode = WAL; 
synchronous = NORMAL; 
busy_timeout = 30000;
We are not at all setting the PRAGMA for cipher_page_size. It is using the default cipher_page_size of 4096 while creating the db instance.

### Steps to Reproduce
There is no issues when the usage of our app is regular. If we are not using the app for a long time or in some other random cases we got a looping situation printing the following errors:
2024-08-20 13:53:43.618: sqlcipher_page_cipher: hmac check failed for pgno=1
2024-08-20 13:53:43.618: sqlite3Codec: error decrypting page 1 data: 1
2024-08-20 13:53:43.618: sqlcipher_codec_ctx_set_error 1
2024-08-20 13:53:43.624: sqlcipher_page_cipher: hmac check failed for pgno=1
2024-08-20 13:53:43.624: sqlite3Codec: error decrypting page 1 data: 1
2024-08-20 13:53:43.624: sqlcipher_codec_ctx_set_error 1
2024-08-20 13:53:43.625: sqlcipher_page_cipher: hmac check failed for pgno=1
2024-08-20 13:53:43.625: sqlite3Codec: error decrypting page 1 data: 1
2024-08-20 13:53:43.625: sqlcipher_codec_ctx_set_error 1
2024-08-20 13:53:43.625: sqlcipher_page_cipher: hmac check failed for pgno=1
2024-08-20 13:53:43.625: sqlite3Codec: error decrypting page 1 data: 1
2024-08-20 13:53:43.625: sqlcipher_codec_ctx_set_error 1
2024-08-20 13:53:43.625: sqlcipher_page_cipher: hmac check failed for pgno=1
2024-08-20 13:53:43.625: sqlite3Codec: error decrypting page 1 data: 1
2024-08-20 13:53:43.625: sqlcipher_codec_ctx_set_error 1

After a long time, the app gets crashed.

Upon analysing, we found out that the cipher_page_size is automatically changed into 4064. (As mentioned earlier we are not at all using the PRAGMA to change the cipher_page_size anywhere in the app.)

Any help regarding this issue would be helpful.

SQLCipher version: SQLite ** version 3.45.3

[Crash.txt](https://github.com/user-attachments/files/16675974/Crash.txt)

**Comment by sjlombardo** [comment]
@Rozariozaro what specifically leads you to believe that the cipher page size is changing? I would note that oyu are setting cipher_plaintext_header_size = 32. That would cause SQLCipher to treat the first page as if it is 4064 bytes (4096-32).

**Comment by Rozariozaro** [comment]
We are recently facing this loop of errors saying hmac failed errors randomly. 
I created a db with cipher_page_size as 8192. And  after loading the data, I quit the app and opened the db instance with cipher_page_size as 4096. And this looping of error statements can be reproduced every time. 
on checking pragma cipher_integrity_check the error message says that hmac failed.

**Comment by sjlombardo** [comment]
Hello @Rozariozaro as far as I can tell from your description that is the expected behavior. SQLCipher does not automatically detect non-standard page sizes. If you create a database with a non-standard page size like 8192, then it is the responsibility of any other application attempting to open the database to call `PRAGMA cipher_page_size = 8192` after keying and prior to accessing the database content. If you fail to do so, SQLCipher will assume the page size is the default of 4096, and any attempts to access the database will fail, and integrity check will likewise report errors. 

In summary, if you change the cipher_page_size or cipher_plaintext_header_size your application will need to specify the appropriate settings every time the database is open.

**Comment by Rozariozaro** [comment]
@sjlombardo we are getting the same kind of the errors “hmac check failed” in our app randomly. We are explicitly setting the cipher_plaintext_header_size = 32,
cipher_salt = “salt_key”, 
journal_mode = wal, 
while opening the database. 

I am just saying that I can reproduce the same kind of errors only while following the above mentioned steps.

Is there any other issue that needs to be checked from our side? 
Is any other information needed to identify the cause?

**Comment by sjlombardo** [comment]
Hello @Rozariozaro - 

The errors you posted are typically caused when an application attempts to open a SQLCipher database with an incorrect encryption key, no key at all, or different settings than should be used. Have you checked all your code paths?

Are you getting actual exceptions / errors in your application, or are you just seeing error messages in the logs / console? 
If the former, what is the application doing at the time the errors occur?
How are you obtaining the key material passed to sqlite3_key_v2 or PRAGMA key?

**Comment by Rozariozaro** [comment]
We are using a singleton class to open, key, apply pragma settings and pass the db instance.
We are getting these logs in console. These 3 logs are constantly been posted on the console (looped maybe). 
Obtained output:
1. The application hangs on the launch screen and then crashes.
2. The application goes to the root view and on any fetch, insert, update, delete calls everything is failing.

Note: This does not happen often. Only on some random occasions that we are unable to identify.
> sqlite3_open
we are using this to open the db, then
> sqlite3_key
we are using this to key the db, 
then applying these pragma settings using 
> sqlite3_exec

PRAGMA cipher_plaintext_header_size = 32;
PRAGMA cipher_salt = "cipher_salt";
PRAGMA journal_mode = WAL;


Should we use the "sqlite3_key_v2" to key the db?

**Comment by sjlombardo** [comment]
Hello @Rozariozaro using sqlite3_key_v2 should not matter.  I do not see obvious problems with the usage you are describing. The most likely scenario is that somehow invalid key material is being provided to SQLCipher. 

Are you able to create a small standalone application that reproduces this issue? If so, share it on github and we would be happy to take a look.

**Comment by sjlombardo** [comment]
Closing for lack of updates.
